### DISCUSSION TITLE
Fetch & set variables within a scenario

### DISCUSSION THREAD

**Discussion by utkarshdeep** [artilleryio/artillery#3231] [Artillery]
Hi all,
Sadly I can't share the exact artillery code due to restrictions on sharing policy imposed by organization but will try to explain the scenario as much as I can.

So I have 1 load test scenario, that is run under the load of 35 reqs per min. Under this scenario, there are 4 sequential API requests. What I want is: based on 3rd API response, I want to look up in a JSON file this property for its corresponding value (think of like a key-value pair lookup). After I get the value, I want to use that as part of the 4th API (POST) request body. How to implement this?

This might be straightforward but have been struggling badly at this. 

Also is there a way to disable metrics-by-endpoint in the HTML report, at it makes the HTML size too much to load in the browser.

**Comment by bernardobridge** [comment]
Hello @utkarshdeep 

You can load custom code using a [processor function](https://www.artillery.io/docs/reference/test-script#processor---load-custom-code). I suggest for this you use an `afterResponse` [hook](https://www.artillery.io/docs/reference/engines/http#afterresponse-hooks) to do that using NodeJS. So basically something like:

```js
const myJson = require('path_to_json.json');

function myAfterResponseHandler(req, res, context, ee, next) {
    //do any lookup you need in the json here
   // the response object will be available as the second arg to this function
    next();
};

module.exports = { myAfterResponseHandler }
```

And hook it in your request like this:

```yml
- post:
    url: '/some/route'
    afterResponse: 'myAfterResponseHandler'
```

You'll also need to make sure you set the processor file in your config (see the docs I linked above).

**Reply by utkarshdeep** [reply]
Can I set some variable in `afterResponse` hook of previous API call & access those variables in the `beforeRequest` hook of next API call?

**Reply by bernardobridge** [reply]
Yes, you can use `context.vars` to set them. For example:

```js
const myJson = require('path_to_json.json');

function myAfterResponseHandler(req, res, context, ee, next) {
    //do any lookup you need in the json here
   // the response object will be available as the second arg to this function
    context.vars.myVariable = '12345'
    next();
};

module.exports = { myAfterResponseHandler }
```

**Comment by bernardobridge** [comment]
>Also is there a way to disable metrics-by-endpoint in the HTML report, at it makes the HTML size too much to load in the browser.

For this, I would suggest you group your url's by name, as this is possibly happening due to too many dynamic urls. Here is an example on how to do it:
https://www.artillery.io/docs/reference/extensions/metrics-by-endpoint#useonlyrequestnames-example
